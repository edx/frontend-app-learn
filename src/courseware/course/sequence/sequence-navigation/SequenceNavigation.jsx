import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@edx/paragon';
import classNames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import { useSelector } from 'react-redux';
import UnitButton from './UnitButton';
import SequenceNavigationTabs from './SequenceNavigationTabs';
import { useSequenceNavigationMetadata } from './hooks';
import { useModel } from '../../../../model-store';

export default function SequenceNavigation({
  unitId,
  sequenceId,
  className,
  onNavigate,
  nextSequenceHandler,
  previousSequenceHandler,
}) {
  const sequence = useModel('sequences', sequenceId);
  const { isFirstUnit, isLastUnit } = useSequenceNavigationMetadata(sequenceId, unitId);
  const isLocked = sequence.gatedContent !== undefined && sequence.gatedContent.gated;
  const sequenceStatus = useSelector(state => state.courseware.sequenceStatus);

  const renderUnitButtons = () => {
    if (isLocked) {
      return (
        <UnitButton unitId={unitId} title="" icon="lock" isActive onClick={() => {}} />
      );
    }
    if (sequence.unitIds.length === 0 || unitId === null) {
      return (
        <div style={{ flexBasis: '100%', minWidth: 0, borderBottom: 'solid 1px #EAEAEA' }} />
      );
    }
    return (
      <SequenceNavigationTabs
        unitIds={sequence.unitIds}
        unitId={unitId}
        showCompletion={sequence.showCompletion}
        onNavigate={onNavigate}
      />
    );
  };

  return sequenceStatus === 'loaded' && (
    <nav className={classNames('sequence-navigation', className)}>
      <Button className="previous-btn" onClick={previousSequenceHandler} disabled={isFirstUnit}>
        <FontAwesomeIcon icon={faChevronLeft} className="mr-2" size="sm" />
        <FormattedMessage
          defaultMessage="Previous"
          id="learn.sequence.navigation.previous.button"
          description="The Previous button in the sequence nav"
        />
      </Button>
      {renderUnitButtons()}
      <Button className="next-btn" onClick={nextSequenceHandler} disabled={isLastUnit}>
        <FormattedMessage
          defaultMessage="Next"
          id="learn.sequence.navigation.next.button"
          description="The Next button in the sequence nav"
        />
        <FontAwesomeIcon icon={faChevronRight} className="ml-2" size="sm" />
      </Button>
    </nav>
  );
}

SequenceNavigation.propTypes = {
  sequenceId: PropTypes.string.isRequired,
  unitId: PropTypes.string,
  className: PropTypes.string,
  onNavigate: PropTypes.func.isRequired,
  nextSequenceHandler: PropTypes.func.isRequired,
  previousSequenceHandler: PropTypes.func.isRequired,
};

SequenceNavigation.defaultProps = {
  className: null,
  unitId: null,
};
