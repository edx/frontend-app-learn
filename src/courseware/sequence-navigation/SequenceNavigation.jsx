import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@edx/paragon';
import classNames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import UnitButton from './UnitButton';
import SequenceNavigationTabs from './SequenceNavigationTabs';
import { useSequenceNavigationMetadata } from './hooks';
import { useModel } from '../../model-store';

export default function SequenceNavigation({
  unitId,
  sequenceId,
  className,
  onNavigate,
  nextSequenceHandler,
  previousSequenceHandler,
}) {
  const sequence = useModel('sequences', sequenceId);
  const { isFirstUnit, isLastUnit } = useSequenceNavigationMetadata(sequenceId, unitId);
  const isLocked = sequence.gatedContent !== undefined && sequence.gatedContent.gated;

  return (
    <nav className={classNames('sequence-navigation', className)}>
      <Button className="previous-btn" onClick={previousSequenceHandler} disabled={isFirstUnit}>
        <FontAwesomeIcon icon={faChevronLeft} className="mr-2" size="sm" />
        <FormattedMessage
          defaultMessage="Previous"
          id="learn.sequence.navigation.previous.button"
          description="The Previous button in the sequence nav"
        />
      </Button>

      {isLocked ? <UnitButton unitId={unitId} title="" contentType="lock" isActive onClick={() => {}} /> : (
        <SequenceNavigationTabs
          unitIds={sequence.unitIds}
          unitId={unitId}
          showCompletion={sequence.showCompletion}
          onNavigate={onNavigate}
        />
      )}

      <Button className="next-btn" onClick={nextSequenceHandler} disabled={isLastUnit}>
        <FormattedMessage
          defaultMessage="Next"
          id="learn.sequence.navigation.next.button"
          description="The Next button in the sequence nav"
        />
        <FontAwesomeIcon icon={faChevronRight} className="ml-2" size="sm" />
      </Button>
    </nav>
  );
}

SequenceNavigation.propTypes = {
  unitId: PropTypes.string.isRequired,
  sequenceId: PropTypes.string.isRequired,
  className: PropTypes.string,
  onNavigate: PropTypes.func.isRequired,
  nextSequenceHandler: PropTypes.func.isRequired,
  previousSequenceHandler: PropTypes.func.isRequired,
};

SequenceNavigation.defaultProps = {
  className: null,
};
