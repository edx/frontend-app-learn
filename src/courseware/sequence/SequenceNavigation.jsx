import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@edx/paragon';
import classNames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import UnitButton from './UnitButton';
import SequenceNavigationTabs from './SequenceNavigationTabs';

export default function SequenceNavigation({
  activeUnitId,
  className,
  isFirstUnit,
  isLastUnit,
  isLocked,
  onNavigate,
  onNext,
  onPrevious,
  showCompletion,
  unitIds,
}) {
  return (
    <nav className={classNames('sequence-navigation', className)}>
      <Button className="previous-btn" onClick={onPrevious} disabled={isFirstUnit}>
        <FontAwesomeIcon icon={faChevronLeft} className="mr-2" size="sm" />
        <FormattedMessage
          defaultMessage="Previous"
          id="learn.sequence.navigation.previous.button"
          description="The Previous button in the sequence nav"
        />
      </Button>


      {isLocked ? <UnitButton type="lock" isActive /> : (
        <SequenceNavigationTabs
          unitIds={unitIds}
          activeUnitId={activeUnitId}
          showCompletion={showCompletion}
          onNavigate={onNavigate}
        />
      )}

      <Button className="next-btn" onClick={onNext} disabled={isLastUnit}>
        <FormattedMessage
          defaultMessage="Next"
          id="learn.sequence.navigation.next.button"
          description="The Next button in the sequence nav"
        />
        <FontAwesomeIcon icon={faChevronRight} className="ml-2" size="sm" />
      </Button>
    </nav>
  );
}

SequenceNavigation.propTypes = {
  activeUnitId: PropTypes.string.isRequired,
  className: PropTypes.string,
  isFirstUnit: PropTypes.bool.isRequired,
  isLastUnit: PropTypes.bool.isRequired,
  isLocked: PropTypes.bool.isRequired,
  onNavigate: PropTypes.func.isRequired,
  onNext: PropTypes.func.isRequired,
  onPrevious: PropTypes.func.isRequired,
  showCompletion: PropTypes.bool.isRequired,
  unitIds: PropTypes.arrayOf(PropTypes.string).isRequired,
};

SequenceNavigation.defaultProps = {
  className: null,
};
