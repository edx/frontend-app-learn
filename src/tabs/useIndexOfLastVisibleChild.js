import { useLayoutEffect, useRef, useState } from 'react';
import useWindowSize from './useWindowSize';

const invisibleStyle = {
  position: 'absolute',
  left: 0,
  pointerEvents: 'none',
  visibility: 'hidden',
};

export default function useIndexOfLastVisibleChild() {
  const containerElementRef = useRef(null);
  const overflowElementRef = useRef(null);
  const containingRectRef = useRef({});
  const [indexOfLastVisibleChild, setIndexOfLastVisibleChild] = useState(-1);
  const windowSize = useWindowSize();

  useLayoutEffect(() => {
    const containingRect = containerElementRef.current.getBoundingClientRect();

    // No-op if the width is unchanged.
    // (Assumes tabs themselves don't change count or width).
    if (!containingRect.width === containingRectRef.current.width) {
      return;
    }
    // Update for future comparison
    containingRectRef.current = containingRect;

    // Get array of child nodes from NodeList form
    const childNodesArr = Array.prototype.slice.call(containerElementRef.current.children);
    const { nextIndexOfLastVisibleChild } = childNodesArr
      // filter out the overflow element
      .filter(childNode => childNode !== overflowElementRef.current)
      // sum the widths to find the last visible element's index
      .reduce((acc, childNode, index) => {
        // use floor to prevent rounding errors
        acc.sumWidth += Math.floor(childNode.getBoundingClientRect().width);
        if (acc.sumWidth <= containingRect.width) {
          acc.nextIndexOfLastVisibleChild = index;
        }
        return acc;
      }, {
        // Include the overflow element's width to begin with. Doing this means
        // sometimes we'll show a dropdown with one item in it when it would fit,
        // but allowing this case dramatically simplifies the calculations we need
        // to do above.
        sumWidth: overflowElementRef.current ? overflowElementRef.current.getBoundingClientRect().width : 0,
        nextIndexOfLastVisibleChild: -1,
      });


    setIndexOfLastVisibleChild(nextIndexOfLastVisibleChild);
  }, [windowSize, containerElementRef.current]);

  return [indexOfLastVisibleChild, containerElementRef, invisibleStyle, overflowElementRef];
}
