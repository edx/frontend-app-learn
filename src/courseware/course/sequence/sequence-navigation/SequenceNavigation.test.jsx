import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { cloneDeep } from 'lodash';
import { fireEvent, getByText } from '@testing-library/dom';
import {
  initialState, render, screen, testUnits,
} from '../../../../setupTest';
import SequenceNavigation from './SequenceNavigation';
import useIndexOfLastVisibleChild from '../../../../generic/tabs/useIndexOfLastVisibleChild';

// Mock the hook to avoid relying on its implementation and mocking `getBoundingClientRect`.
jest.mock('../../../../generic/tabs/useIndexOfLastVisibleChild');
useIndexOfLastVisibleChild.mockReturnValue([0, null, null]);

describe('Sequence Navigation', () => {
  const mockData = {
    previousSequenceHandler: () => {},
    onNavigate: () => {},
    nextSequenceHandler: () => {},
    sequenceId: '1',
    unitId: '3',
  };

  it('is empty while loading', () => {
    // Clone initialState.
    const testState = cloneDeep(initialState);
    testState.courseware.sequenceStatus = 'loading';

    const { container } = render(
      <SequenceNavigation {...mockData} />,
      { initialState: testState },
    );
    expect(container).toBeEmptyDOMElement();
  });

  it('renders empty div without unitId', () => {
    const { container } = render(<SequenceNavigation {...mockData} unitId={undefined} />, { initialState });
    expect(getByText(container, (content, element) => (
      element.tagName.toLowerCase() === 'div' && element.getAttribute('style')))).toBeEmptyDOMElement();
  });

  it('renders locked button for gated content', () => {
    const testState = cloneDeep(initialState);
    testState.models.sequences['1'].gatedContent = { gated: true };
    const onNavigate = jest.fn();
    render(<SequenceNavigation {...mockData} {...{ onNavigate }} />, { initialState: testState });

    const unitButton = screen.getByTitle(mockData.unitId);
    fireEvent.click(unitButton);
    // The unit button should not work for gated content.
    expect(onNavigate).not.toHaveBeenCalled();
    // TODO: Not sure if this is working as expected, because the `contentType="lock"` will be overridden by the value
    //  from Redux. To make this provide a `fa-icon` lock we could introduce something like `overriddenContentType`.
    expect(unitButton.firstChild).toHaveClass('fa-book');
  });

  it('renders correctly and handles unit button clicks', () => {
    const onNavigate = jest.fn();
    render(<SequenceNavigation {...mockData} {...{ onNavigate }} />, { initialState });

    const unitButtons = screen.getAllByRole('button', { name: /\d+/ });
    expect(unitButtons).toHaveLength(testUnits.length);
    unitButtons.forEach(button => fireEvent.click(button));
    expect(onNavigate).toHaveBeenCalledTimes(unitButtons.length);
  });

  it('has both navigation buttons enabled for a non-corner unit of the sequence', () => {
    render(<SequenceNavigation {...mockData} />, { initialState });

    screen.getAllByRole('button', { name: /previous|next/i }).forEach(button => {
      expect(button).toBeEnabled();
    });
  });

  it('has the "Previous" button disabled for the first unit of the sequence', () => {
    render(<SequenceNavigation {...mockData} unitId="1" />, { initialState });

    expect(screen.getByRole('button', { name: /previous/i })).toBeDisabled();
    expect(screen.getByRole('button', { name: /next/i })).toBeEnabled();
  });

  it('has the "Next" button disabled for the last unit of the sequence', () => {
    render(<SequenceNavigation
      {...mockData}
      sequenceId="2"
      unitId={testUnits.length.toString()}
    />, { initialState });

    expect(screen.getByRole('button', { name: /previous/i })).toBeEnabled();
    expect(screen.getByRole('button', { name: /next/i })).toBeDisabled();
  });

  it('handles "Previous" and "Next" click', () => {
    const previousSequenceHandler = jest.fn();
    const nextSequenceHandler = jest.fn();
    render(<SequenceNavigation
      {...mockData}
      {...{ previousSequenceHandler, nextSequenceHandler }}
    />, { initialState });

    fireEvent.click(screen.getByRole('button', { name: /previous/i }));
    expect(previousSequenceHandler).toHaveBeenCalledTimes(1);

    fireEvent.click(screen.getByRole('button', { name: /next/i }));
    expect(nextSequenceHandler).toHaveBeenCalledTimes(1);
  });
});
