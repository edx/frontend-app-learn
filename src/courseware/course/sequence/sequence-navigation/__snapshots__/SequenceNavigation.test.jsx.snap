// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequence Navigation is empty while loading 1`] = `<DocumentFragment />`;

exports[`Sequence Navigation renders correctly 1`] = `
<DocumentFragment>
  <nav
    class="sequence-navigation"
  >
    <button
      class="btn previous-btn"
      type="button"
    >
      <img
        alt="fa-chevron-left"
        class="fa-chevron-left"
        data-testid="icon"
      />
      <span>
        Previous
      </span>
    </button>
    <div
      style="flex-basis: 100%; min-width: 0;"
    >
      <div
        class="sequence-navigation-tabs-container"
      >
        <div
          class="sequence-navigation-tabs d-flex flex-grow-1"
        >
          <button
            class="btn"
            title="1"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="2"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn active"
            title="3"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="4"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="5"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="6"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="7"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="8"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="9"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
          <button
            class="btn"
            title="10"
            type="button"
          >
            <img
              alt="fa-book"
              class="fa-book"
              data-testid="icon"
            />
          </button>
        </div>
      </div>
    </div>
    <button
      class="btn next-btn"
      type="button"
    >
      <span>
        Next
      </span>
      <img
        alt="fa-chevron-right"
        class="fa-chevron-right"
        data-testid="icon"
      />
    </button>
  </nav>
</DocumentFragment>
`;

exports[`Sequence Navigation renders empty div without unitId 1`] = `
<DocumentFragment>
  <nav
    class="sequence-navigation"
  >
    <button
      class="btn previous-btn"
      type="button"
    >
      <img
        alt="fa-chevron-left"
        class="fa-chevron-left"
        data-testid="icon"
      />
      <span>
        Previous
      </span>
    </button>
    <div
      style="flex-basis: 100%; min-width: 0; border-bottom: 1px solid #EAEAEA;"
    />
    <button
      class="btn next-btn"
      type="button"
    >
      <span>
        Next
      </span>
      <img
        alt="fa-chevron-right"
        class="fa-chevron-right"
        data-testid="icon"
      />
    </button>
  </nav>
</DocumentFragment>
`;

exports[`Sequence Navigation renders locked button for gated content 1`] = `
<DocumentFragment>
  <nav
    class="sequence-navigation"
  >
    <button
      class="btn previous-btn"
      type="button"
    >
      <img
        alt="fa-chevron-left"
        class="fa-chevron-left"
        data-testid="icon"
      />
      <span>
        Previous
      </span>
    </button>
    <button
      class="btn active"
      title="3"
      type="button"
    >
      <img
        alt="fa-book"
        class="fa-book"
        data-testid="icon"
      />
    </button>
    <button
      class="btn next-btn"
      type="button"
    >
      <span>
        Next
      </span>
      <img
        alt="fa-chevron-right"
        class="fa-chevron-right"
        data-testid="icon"
      />
    </button>
  </nav>
</DocumentFragment>
`;
